File: apps/web/src/modules/offers/services/offersService.js
import api from '../../../services/api';

const API_BASE = '/offers';

export const offersService = {
  getOffersByBusiness: async (businessId) => {
    return api.get(`${API_BASE}/business/${businessId}`);
  },

  getOfferByCode: async (code) => {
    return api.get(`${API_BASE}/code/${code}`);
  },

  getOfferById: async (id) => {
    return api.get(`${API_BASE}/${id}`);
  },

  createOffer: async (businessId, data) => {
    return api.post(`${API_BASE}/business/${businessId}`, data);
  },

  updateOffer: async (id, data) => {
    return api.patch(`${API_BASE}/${id}`, data);
  },

  deleteOffer: async (id) => {
    return api.delete(`${API_BASE}/${id}`);
  }
};

export default offersService;

File: apps/mobile/lib/modules/offers/services/offers_service.dart
import 'package:dio/dio.dart';
import '../../../config/api_client.dart';
import '../models/offer_model.dart';

class OffersService {
  final Dio _dio = ApiClient.getDio();
  static const String _baseUrl = '/api/offers';

  Future<List<OfferModel>> getOffersByBusiness(String businessId) async {
    try {
      final response = await _dio.get(
        _baseUrl,
        queryParameters: {'businessId': businessId},
      );

      if (response.statusCode == 200) {
        final data = response.data;
        final List list = data is List ? data : (data['data'] ?? []);
        return list
            .map((item) => OfferModel.fromJson(item as Map<String, dynamic>))
            .toList();
      }
      throw Exception('Failed to load offers');
    } catch (e) {
      rethrow;
    }
  }

  Future<OfferModel> getOfferById(String id) async {
    try {
      final response = await _dio.get('$_baseUrl/$id');

      if (response.statusCode == 200) {
        return OfferModel.fromJson(response.data as Map<String, dynamic>);
      }
      throw Exception('Failed to load offer');
    } catch (e) {
      rethrow;
    }
  }

  Future<OfferModel> validateCode(String code) async {
    try {
      final response = await _dio.post(
        '$_baseUrl/validate',
        data: {'code': code},
      );

      if (response.statusCode == 200) {
        return OfferModel.fromJson(response.data as Map<String, dynamic>);
      }
      throw Exception('Invalid or expired offer code');
    } catch (e) {
      rethrow;
    }
  }
}
